x/qgb/orchestrator/orchestrator.go

```diff
@@ -4,95 +4,436 @@ import (
+	return nil
+}
+
+var _ Broadcaster = &broadcaster{}
+
+type Broadcaster interface {
+	BroadcastTx(ctx context.Context, msg sdk.Msg) (string, error)
+}
+
+type broadcaster struct {
+	mutex   *sync.Mutex
+	signer  *paytypes.KeyringSigner
+	qgbGrpc *grpc.ClientConn
+}
+
+func NewBroadcaster(qgbGrpcAddr string, signer *paytypes.KeyringSigner) (Broadcaster, error) {
+	qgbGrpc, err := grpc.Dial(qgbGrpcAddr, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	return &broadcaster{
+		mutex:   &sync.Mutex{}, // investigate if this is needed
+		signer:  signer,
+		qgbGrpc: qgbGrpc,
+	}, nil
+}
+
+func (bc *broadcaster) BroadcastTx(ctx context.Context, msg sdk.Msg) (string, error) {
+	bc.mutex.Lock()
+	defer bc.mutex.Unlock()
+	err := bc.signer.QueryAccountNumber(ctx, bc.qgbGrpc)
+	if err != nil {
+		return "", err
+	}
+
+	builder := bc.signer.NewTxBuilder()
+	// TODO make gas limit configurable
+	builder.SetGasLimit(9999999999999)
+	// TODO: update this api
+	// via https://github.com/celestiaorg/celestia-app/pull/187/commits/37f96d9af30011736a3e6048bbb35bad6f5b795c
+	tx, err := bc.signer.BuildSignedTx(builder, msg)
+	if err != nil {
+		return "", err
+	}
+
+	rawTx, err := bc.signer.EncodeTx(tx)
+	if err != nil {
+		return "", err
+	}
+
+	// FIXME sdktypestx.BroadcastMode_BROADCAST_MODE_BLOCK waits for a block to be minted containing
+	// the transaction to continue. This makes the orchestrator slow to catchup.
+	// It would be better to just send the transaction. Then, another job would keep an eye
+	// if the transaction was included. If not, retry it. But this would mean we should increment ourselves
+	// the sequence number after each broadcasted transaction.
+	// We can also use BroadcastMode_BROADCAST_MODE_SYNC but it will also fail due to a non incremented
+	// sequence number.
+
+	// TODO  check if we can move this outside of the paytypes
+	resp, err := paytypes.BroadcastTx(ctx, bc.qgbGrpc, sdktypestx.BroadcastMode_BROADCAST_MODE_BLOCK, rawTx)
+	if err != nil {
+		return "", err
+	}
+
+	if resp.TxResponse.Code != 0 {
+		return "", errors.Wrap(ErrFailedBroadcast, resp.TxResponse.RawLog)
+	}
+
+	return resp.TxResponse.TxHash, nil
+}
+
+var _ Retrier = &retrier{}
+
+type retrier struct {
+	logger        tmlog.Logger
+	retriesNumber int
+}
+
+func NewRetrier(logger tmlog.Logger, retriesNumber int) *retrier {
+	return &retrier{
+		logger:        logger,
+		retriesNumber: retriesNumber,
+	}
+}
+
+type Retrier interface {
+	Retry(nonce uint64, ctx context.Context, retryMethod func(context.Context, uint64) error) error
+	RetryThenFail(nonce uint64, ctx context.Context, retryMethod func(context.Context, uint64) error)
+}
+
+func (r retrier) Retry(nonce uint64, ctx context.Context, retryMethod func(context.Context, uint64) error) error {
+	var err error
+	for i := 0; i <= r.retriesNumber; i++ {
+		// We can implement some exponential backoff in here
```

Comment:
@evan-forbes do we want to have this in this PR? or we leave it for later?

Reply:
we can leave it for later